chain/chain/src/store.rs

```diff
@@ -234,6 +328,74 @@ impl ChainStore {
             }
         }
     }
+
+    pub fn check_blocks_on_same_chain(
+        &mut self,
+        cur_header: &BlockHeader,
+        base_block_hash: &CryptoHash,
+        max_difference_in_height: u64,
+    ) -> bool {
+        // first step: update cache head
+        if self.header_history.is_empty() {
+            self.header_history.push_back(cur_header.clone());
+        }
+        let mut prev_block_hash = cur_header.inner.prev_hash;
+
+        let contains_hash = self.header_history.update(&cur_header.hash, vec![]);
+        if !contains_hash {
+            let mut header_list = vec![cur_header.clone()];
+            let mut find_ancestor = false;
+            while !self.header_history.is_empty() {
+                let prev_block_header = if let Ok(header) = self.get_block_header(&prev_block_hash)
+                {
+                    header.clone()
+                } else {
+                    return false;
+                };
+                self.header_history.pop_front();
+                if self.header_history.update(&prev_block_header.hash, header_list.clone()) {
```

Comment:
Cloning the headerlist every time makes the complexity here quadratic.
Consider my comment above